### PR TITLE
react-dom package update with test-utils

### PR DESCRIPTION
**Pull request by MartynasZilinskas**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [ReactTestUtils](https://facebook.github.io/react/docs/test-utils.html)
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "../tslint.json" }`.

---------------------------------------------------------------------------

Changes:
* `react-addons-test-utils` package moved to `react-dom/test-utils` [ReactTestUtils doc](https://facebook.github.io/react/docs/test-utils.html)
* Added tests
* Added jsdocs comments from documentation
* Fixed some inconsistency with current documentation: 
    * `ReactTestUtils.isCompositeComponentWithType()` returns a boolean [Docs reference](https://facebook.github.io/react/docs/test-utils.html#iscompositecomponentwithtype) - (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15732/commits/7823070f84a88c681632260c39c5d89ebe70d20f#diff-fdca647c016cb4d2f2709c0b8a0669f9R260)
   * Added missing function `ReactTestUtils.isElement()` [Docs reference](https://facebook.github.io/react/docs/test-utils.html#iselement) - (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15732/commits/438ad2e1788c8207837dbfdfa9dca8e9a836b2c1#diff-fdca647c016cb4d2f2709c0b8a0669f9R183)

---------------------------------------------------------------------------

~~I had to disable tslint rule `export-just-namespace`.~~ 
```json
{
    "extends": "dtslint/dt.json",
    "rules": {
        "export-just-namespace": false
    }
}
```
Because it returns itslef (ReactTestUtils object) for chaining. 
[Link to source](https://github.com/facebook/react/blob/master/src/test/ReactTestUtils.js#L356-L364) 
[Link to types commit](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15732/files#diff-fdca647c016cb4d2f2709c0b8a0669f9R165)

**EDIT:**
Solution for this is to use `import * as ReactTestUtils from ".";`

---------------------------------------------------------------------------

Previous PR: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15732

Doc links:
[ReactDOM](https://facebook.github.io/react/docs/react-dom.html)
[ReactDOM server](https://facebook.github.io/react/docs/react-dom-server.html)
[ReactTestUtils](https://facebook.github.io/react/docs/test-utils.html)

### CODE REVIEW

**Comment by andy-ms** [comment]
Sorry, like travis says, we're using `dtslint/dt.json` now instead of `../tslint.json`. You may want to `rm -r node_modules; npm install` too.
As for `export-just-namespace`: Add `import * as ReactTestUtils from "."; // Module returns itself for chaining`, delete `export = ReactTestUtils`, remove the `declare namespace ReactTestUtils { }` wrapper, unindent all code, and add the `export` keyword to everything.

**Comment by MartynasZilinskas** [comment]
Hey @andy-ms, thanks for fast response. Looks like I fixed remaining errors.

**Comment by DovydasNavickas** [comment]
@andy-ms, our test pipelines throw warning because of this PR not being merged. Could you review it?

File: types/react-dom/index.d.ts [hunk 1]
```diff
@@ -1,63 +1,65 @@
-// Type definitions for React (react-dom) 0.14
+// Type definitions for React (react-dom) 15.5
 // Project: http://facebook.github.io/react/
-// Definitions by: Asana <https://asana.com>, AssureSign <http://www.assuresign.com>, Microsoft <https://microsoft.com>
+// Definitions by: Asana <https://asana.com>, AssureSign <http://www.assuresign.com>, Microsoft <https://microsoft.com>, MartynasZilinskas <https://github.com/MartynasZilinskas>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 // TypeScript Version: 2.1
 
-export as namespace ReactDOM;
```
Comment (andy-ms) [inline types/react-dom/index.d.ts]: This line was removed -- is the package no longer available as a global variable `ReactDOM`?

File: types/react-dom/test-utils/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+export function scryRenderedComponentsWithType<T extends Component<{}, {}>, C extends ComponentClass<{}>>(
+    root: Component<any, any>,
+    type: ClassType<any, T, C>): T[];
+
+/**
+ * Same as `scryRenderedComponentsWithType()` but expects there to be one result
+ * and returns that one result, or throws exception if there is any other
+ * number of matches besides one.
+ *
+ * @export
```
Comment (andy-ms) [inline types/react-dom/test-utils/index.d.ts]: This (`@export`, `@template`, `@param` with no description, `@returns` with no description) just duplicates what's in the definition already.

Comment (MartynasZilinskas) [inline types/react-dom/test-utils/index.d.ts]: Yeah, removed all JSDocs tags that has no description.

File: types/react-dom/index.d.ts (hunk 1 above)
Comment (MartynasZilinskas) [inline types/react-dom/index.d.ts]: My bad, reverted removed line.

**Comment by andy-ms** [comment]
Thanks!

**Comment by rebornix** [comment]
@MartynasZilinskas @andy-ms I'm not sure if I made anything with setup, but I kept getting 

> [14:35:08] Error: /Users/penlv/code/vscode/node_modules/@types/react-dom/test-utils/index.d.ts(8,33): Cannot find module '.'

any idea about this error?

**Comment by andy-ms** [comment]
@rebornix What TypeScript version are you using? What are your compiler options? That self-import *should* work in any module resolution setting.

**Comment by MartynasZilinskas** [comment]
Hey @rebornix, I have been using with latest TypeScript version and I haven't encountered any errors.

**Pull request merged**